File: Templates/Matrix/matrix.go
package liner

import (
	"errors"
	"fmt"
	"math/big"
	"strconv"
	"strings"
)

type Mat struct {
	data []big.Rat
	row  int
	col  int
}

func NewMatRat(inputData []big.Rat, m int, n int) *Mat {
	if m < 0 || n < 0 || m*n != len(inputData) {
		panic("error size for the matrix")
	}
	data := make([]big.Rat, len(inputData))
	for i := 0; i < len(data); i++ {
		data[i].Set(&inputData[i])
	}
	return &Mat{data: data, row: m, col: n}
}

func NewMat(data []int64, m int, n int) *Mat {
	if m < 0 || n < 0 || m*n != len(data) {
		panic("error size for the matrix")
	}
	ratData := make([]big.Rat, len(data))
	for i := 0; i < len(data); i++ {
		ratData[i].SetInt64(data[i])
	}
	return &Mat{data: ratData, row: m, col: n}
}

func New1Mat(m int, n int) *Mat {
	ratData := make([]big.Rat, m*n)
	if m <= 0 || n <= 0 {
		panic("error size for the matrix")
	}
	for i := 0; i < m*n; i++ {
		ratData[i].SetInt64(1)
	}
	return &Mat{data: ratData, row: m, col: n}
}

func (inputMat *Mat) String() (s string) {
	format := "%"
	max := 0
	for _, k := range inputMat.data {
		l := len(k.String())
		if k.IsInt() {
			l -= 2
		}
		if l > max {

			max = l
		}
	}
	format += strconv.FormatInt(int64(max), 10)

	/* First line */
	s += "┌ "
	s += strings.Repeat(" ", inputMat.col*(max+1))
	s += "┐\n"

	/* For each element */
	for i := 0; i < inputMat.row; i++ {
		s += "│ "
		for j := 0; j < inputMat.col; j++ {
			if inputMat.data[i*inputMat.col+j].IsInt() {
				s += fmt.Sprintf(format+"d ", inputMat.data[i*inputMat.col+j].Num())
			} else {
				s += fmt.Sprintf(format+"s ", inputMat.data[i*inputMat.col+j].String())
			}
		}
		s += "│\n"
	}

	/* Last line */
	s += "└ "
	s += strings.Repeat(" ", inputMat.col*(max+1))
	s += "┘"
	return s
}

func (inputMat *Mat) Copy() *Mat {
	mat := NewMatRat(inputMat.data, inputMat.row, inputMat.col)
	return mat
}

func (inputMat *Mat) Set(mat *Mat) *Mat {
	data, m, n := mat.Data()
	inputMat.data = data
	inputMat.row = m
	inputMat.col = n
	return inputMat
}

func (inputMat *Mat) Data() ([]big.Rat, int, int) {
	newMat := inputMat.Copy()
	return newMat.data, newMat.row, newMat.col
}

func (x *Mat) Mul(a *Mat, b *Mat) *Mat {
	if a.row == 0 || a.col == 0 || b.row == 0 || b.col == 0 {
		panic("cannot multiply empty matrix")
	}
	if a.col != b.row {
		panic("error size for the matrix to multiply")
	}
	data := make([]big.Rat, a.row*b.col)

	for i := 0; i < a.row; i++ {
		for j := 0; j < b.col; j++ {
			for k := 0; k < a.col; k++ {
				data[i*a.col+j].Add(&data[i*a.col+j], new(big.Rat).Mul(&a.data[i*a.col+k], &b.data[k*b.col+j]))
			}
		}
	}

	copy(x.data, data)

	x.row = a.row
	x.col = b.col
	return x
}

func (x *Mat) Add(a *Mat, b *Mat) *Mat {
	if a.row == 0 || a.col == 0 || b.row == 0 || b.col == 0 {
		panic("cannot multiply empty matrix")
	}
	if a.row != b.row || a.col != b.col {
		panic("error size for the matrix to add")
	}
	x.data = make([]big.Rat, len(a.data))
	x.row = a.row
	x.col = b.col

	for i := 0; i < a.row; i++ {
		for j := 0; j < a.col; j++ {
			x.data[i*a.col+j].Add(&a.data[i*a.col+j], &b.data[i*a.col+j])
		}
	}
	return x
}

func (x *Mat) Neg(a *Mat) *Mat {
	x.Set(a)
	for i := 0; i < len(x.data); i++ {
		x.data[i].Neg(&x.data[i])
	}
	return x
}

func (x *Mat) swapRows(a int, b int) *Mat {
	for i := 0; i < x.col; i++ {
		var temp big.Rat
		temp.Set(&x.data[a*x.col+i])
		x.data[a*x.col+i].Set(&x.data[b*x.col+i])
		x.data[b*x.col+i].Set(&temp)
	}
	return x
}

func (c *Mat) Det(step bool) *big.Rat {
	a := new(Mat).Set(c)
	if step {
		fmt.Println("Input:")
		fmt.Println(a)
	}
	ans := big.NewRat(1, 1)
	if a.row <= 0 || a.col <= 0 {
		panic("cannot calculate determinate of an empty matrix")
	}
	if a.row != a.col {
		panic("not a square")
	}
	for i := 1; i < a.row; i++ {
		l := i - 1
		for ; l < a.row && a.data[l*a.col+i-1].Num().Int64() == 0; l++ {
		}
		if l == a.row {
			return new(big.Rat).SetInt64(0)
		} else if l != i-1 {
			ans.Neg(ans)
			a.swapRows(l, i-1)
			if step {
				fmt.Print("r", l+1, " <-> ", "r", i, "\n")
				fmt.Println(a)
			}
		}
		var inv big.Rat
		inv.Inv(&a.data[(i-1)*a.col+i-1])
		for k := i; k < a.row; k++ {
			numToMul := new(big.Rat).Set(&inv)
			numToMul.Mul(numToMul, &a.data[k*a.col+i-1])
			numToMul.Neg(numToMul)
			if step {
				fmt.Print("r", k+1, " + (", f(numToMul), ")r", i, "\n")
			}
			for j := 0; j < a.col; j++ {
				a.data[k*a.col+j].Add(&a.data[k*a.col+j],
					new(big.Rat).Mul(&a.data[(i-1)*a.col+j], numToMul))
			}
		}
		if step {
			fmt.Println(a)
		}
	}
	for i := 0; i < a.row; i++ {
		ans.Mul(ans, &a.data[i*a.row+i])
	}
	if step {
		fmt.Println("Result:")
		if ans.IsInt() {
			fmt.Println(ans.Num())
		} else {
			fmt.Println(ans)
		}
	}
	return ans
}

func (z *Mat) Scan(s fmt.ScanState, ch rune) error {
	m, n := 0, 0
	fmt.Scan(&m, &n)
	if m <= 0 || n <= 0 {
		return errors.New("error size for the matrix")
	}
	data := make([]big.Rat, m*n)
	for i := 0; i < len(data); i++ {
		fmt.Scan(&data[i])
	}

	z.data = data
	z.row = m
	z.col = n
	return nil
}

func (a *Mat) Row() int {
	return a.row
}

func (a *Mat) Col() int {
	return a.col
}

func Pow(a *Mat, n uint64) *Mat {
	var b Mat
	b.Set(a)
	switch n {
	case 0:
		return New1Mat(a.Row(), a.Col())
	case 1:
		return &b
	default:
		half := Pow(a, n/2)
		half.Mul(half, half)
		if n%2 == 1 {
			return half.Mul(half, a)
		} else {
			return half
		}
	}
}

func (a *Mat) Transpose(b *Mat) *Mat {
	data, m, n := b.Data()
	a.row = n
	a.col = m
	tData := make([]big.Rat, m*n)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			tData[j*m+i].Set(&data[i*n+j])
		}
	}
	a.data = tData
	return a
}

func (a *Mat) Inv(b *Mat, step bool) *Mat {
	if b.row <= 0 || b.col <= 0 {
		panic("cannot calculate determinate of an empty matrix")
	}
	if b.row != b.col {
		panic("not mat square")
	}
	det := b.Det(false)
	if det.Num().Int64() == 0 {
		panic("the determinant is 0")
	}
	var mat Mat
	data := make([]big.Rat, b.col*b.row*2)
	mat.row = b.row
	mat.col = 2 * b.col
	mat.data = data
	for r := 0; r < b.row; r++ {
		for c := 0; c < b.col; c++ {
			data[r*mat.col+c].Set(&b.data[r*b.col+c])
		}
	}
	for r := 0; r < b.row; r++ {
		data[(1+r*2)*b.col+r].Set(new(big.Rat).SetInt64(1))
	}
	if step {
		fmt.Println("Let's assume the new Matrix B")
		fmt.Println(&mat)
	}

	for i := 1; i < mat.row; i++ {
		l := i - 1
		for ; l < mat.row && mat.data[l*mat.col+i-1].Num().Int64() == 0; l++ {
		}
		if l != i-1 {
			mat.swapRows(l, i-1)
			if step {
				fmt.Print("r", l+1, " <-> ", "r", i, "\n")
				fmt.Println(&mat)
			}
		}
		var inv big.Rat
		inv.Inv(&mat.data[(i-1)*mat.col+i-1])
		for k := i; k < mat.row; k++ {
			numToMul := new(big.Rat).Set(&inv)
			numToMul.Mul(numToMul, &mat.data[k*mat.col+i-1])
			numToMul.Neg(numToMul)
			if step {
				fmt.Print("r", k+1, " + (", f(numToMul), ")r", i, "\n")
			}
			for j := 0; j < mat.col; j++ {
				mat.data[k*mat.col+j].Add(&mat.data[k*mat.col+j],
					new(big.Rat).Mul(&mat.data[(i-1)*mat.col+j], numToMul))
			}
		}
		if step {
			fmt.Println(&mat)
		}
	}

	for i := mat.row - 2; i >= 0; i-- {
		var inv big.Rat
		inv.Inv(&mat.data[(i+1)*mat.col+i+1])
		for k := 0; k <= i; k++ {
			numToMul := new(big.Rat).Set(&inv)
			numToMul.Mul(numToMul, &mat.data[k*mat.col+i+1])
			numToMul.Neg(numToMul)
			if step {
				fmt.Print("r", k+1, " + (", f(numToMul), ")r", i+2, "\n")
			}
			for j := 0; j < mat.col; j++ {
				mat.data[k*mat.col+j].Add(&mat.data[k*mat.col+j],
					new(big.Rat).Mul(&mat.data[(i+1)*mat.col+j], numToMul))
			}
		}
		if step {
			fmt.Println(&mat)
		}
	}

	for i := 0; i < mat.row; i++ {
		if data[i*mat.col+i].Cmp(new(big.Rat).SetInt64(1)) != 0 {
			var inv big.Rat
			inv.Inv(&data[i*mat.col+i])
			if step {
				fmt.Print("r", i+1, " x (", f(&inv), ")\n")
			}
			for j := 0; j < mat.col; j++ {
				data[i*mat.col+j].Mul(&data[i*mat.col+j], &inv)
			}
		}
	}

	if step {
		fmt.Println(&mat)
	}

	a.col = b.col
	a.row = b.row
	aData := make([]big.Rat, b.col*b.row)
	a.data = aData
	for i := 0; i < b.row; i++ {
		for j := 0; j < b.col; j++ {
			aData[i*b.col+j] = data[(i*2+1)*b.col+j]
		}
	}
	return a
}

func (a *Mat) Inv2(b *Mat, c *Mat, step bool) *Mat { // returns b^-1 * c
	if b.row <= 0 || b.col <= 0 {
		panic("cannot calculate determinate of an empty matrix")
	}
	if b.row != b.col || b.row != c.row {
		panic("not mat square")
	}
	if b.Det(false).Num().Int64() == 0 {
		panic("the determinant is 0")
	}
	var mat Mat
	data := make([]big.Rat, (c.col+b.col)*b.row)
	mat.row = b.row
	mat.col = c.col + b.col
	mat.data = data
	for r := 0; r < b.row; r++ {
		for c := 0; c < b.col; c++ {
			data[r*mat.col+c].Set(&b.data[r*b.col+c])
		}
	}

	for r := 0; r < b.row; r++ {
		for col := 0; col < c.col; col++ {
			data[r*mat.col+b.col+col].Set(&c.data[r*c.col+col])
		}
	}

	if step {
		fmt.Println("Let's assume the new Matrix B")
		fmt.Println(&mat)
	}

	for i := 1; i < mat.row; i++ {
		l := i - 1
		for ; l < mat.row && mat.data[l*mat.col+i-1].Num().Int64() == 0; l++ {
		}
		if l != i-1 {
			mat.swapRows(l, i-1)
			if step {
				fmt.Print("r", l+1, " <-> ", "r", i, "\n")
				fmt.Println(mat)
			}
		}
		var inv big.Rat
		inv.Inv(&mat.data[(i-1)*mat.col+i-1])
		for k := i; k < mat.row; k++ {
			numToMul := new(big.Rat).Set(&inv)
			numToMul.Mul(numToMul, &mat.data[k*mat.col+i-1])
			numToMul.Neg(numToMul)
			if step {
				fmt.Print("r", k+1, " + (", f(numToMul), ")r", i, "\n")
			}
			for j := 0; j < mat.col; j++ {
				mat.data[k*mat.col+j].Add(&mat.data[k*mat.col+j],
					new(big.Rat).Mul(&mat.data[(i-1)*mat.col+j], numToMul))
			}
		}
		if step {
			fmt.Println(&mat)
		}
	}

	for i := mat.row - 2; i >= 0; i-- {
		var inv big.Rat
		inv.Inv(&mat.data[(i+1)*mat.col+i+1])
		for k := 0; k <= i; k++ {
			numToMul := new(big.Rat).Set(&inv)
			numToMul.Mul(numToMul, &mat.data[k*mat.col+i+1])
			numToMul.Neg(numToMul)
			if step {
				fmt.Print("r", k+1, " + (", f(numToMul), ")r", i, "\n")
			}
			for j := 0; j < mat.col; j++ {
				mat.data[k*mat.col+j].Add(&mat.data[k*mat.col+j],
					new(big.Rat).Mul(&mat.data[(i+1)*mat.col+j], numToMul))
			}
		}
		if step {
			fmt.Println(&mat)
		}
	}

	for i := 0; i < mat.row; i++ {
		if data[i*mat.col+i].Cmp(new(big.Rat).SetInt64(1)) != 0 {
			var inv big.Rat
			inv.Inv(&data[i*mat.col+i])
			if step {
				fmt.Print("r", i+1, " x (", f(&inv), ")\n")
			}
			for j := 0; j < mat.col; j++ {
				data[i*mat.col+j].Mul(&data[i*mat.col+j], &inv)
			}
		}
	}

	if step {
		fmt.Println(&mat)
	}

	a.col = c.col
	a.row = c.row
	aData := make([]big.Rat, c.col*c.row)
	a.data = aData
	for i := 0; i < c.row; i++ {
		for j := 0; j < c.col; j++ {
			aData[i*c.col+j] = data[i*mat.col+b.col+j]
		}
	}
	return a
}

func f(a *big.Rat) interface{} {
	if a.IsInt() {
		return a.Num()
	} else {
		return a
	}
}

func (a *Mat) StandardOutput() {
	fmt.Println(a.row, a.col)
	for i := 0; i < a.row; i++ {
		for j := 0; j < a.col; j++ {
			fmt.Print(f(&a.data[i*a.col+j]))
			if j < a.col-1 {
				fmt.Print(" ")
			} else {
				fmt.Print("\n")
			}
		}
	}
}

File: Templates/Matrix/matrix.cpp
//
// Created by 郭泽宇 on 2019-05-07.
// Using C++11 standard
//

#include "matrix.h"
#include <iostream>
#include <sstream>

using std::cout;
using std::endl;

matrix::matrix(const fraction *d, int r, int c) {
	if (r < 0 || c < 0) {
		throw "error size for the matrix";
	}
	row = r;
	col = c;
	data = new fraction[r * c];
	memcpy(data, d, sizeof(fraction) * r * c);
}

matrix::~matrix() { delete[] data; }

matrix::matrix(const matrix &m) {
	data = new fraction[m.row * m.col];
	row = m.row;
	col = m.col;
	memcpy(data, m.data, sizeof(fraction) * m.row * m.col);
}

// 重载了输出运算符
ostream &operator<<(ostream &os, const matrix &mat1) {
	if (mat1.data == nullptr) {
		os << "[]";
		return os;
	}

	int max = 0;
	for (int i = 0; i < mat1.col * mat1.row; i++) {
		std::ostringstream str;
		{ str << mat1.data[i]; }
		int l = str.str().length();
		if (l > max) {
			max = l;
		}
	}

	// 首行
	os << "┌ ";
	os << std::string(mat1.col * (max + 1), ' ');
	os << "┐\n";

	// 每一个元素
	for (int i = 0; i < mat1.row; ++i) {
		os << "│";
		for (int j = 0; j < mat1.col; ++j) {
			std::ostringstream str;
			// Code found on StackOverflow
			{ str << mat1.data[mat1.p(i, j)]; }
			os << std::string(max + 1 - str.str().length(), ' ');
			os << mat1.data[mat1.p(i, j)];
		}
		os << " │\n";
	}

	// 最后一行
	os << "└ ";
	os << std::string(mat1.col * (max + 1), ' ');
	os << "┘";
	return os;
}

matrix &matrix::operator=(const matrix &m) {
	delete[] data;
	data = new fraction[m.row * m.col];
	row = m.row;
	col = m.col;
	memcpy(data, m.data, sizeof(fraction) * m.row * m.col);
	return *this;
}

matrix matrix::operator+(const matrix &m) const {
	if (col != m.col || row != m.row) {
		throw "error size for the matrix to add";
	}
	if (col == 0 || row == 0) {
		throw "cannot add empty matrix";
	}
	matrix ans;
	ans.data = new fraction[row * col];
	ans.row = row;
	ans.col = col;

	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < col; ++j) {
			ans.data[p(i, j)] = data[p(i, j)] + m.data[p(i, j)];
		}
	}

	return ans;
}

matrix matrix::operator-() const {
	matrix ans;
	ans.data = new fraction[row * col];
	ans.row = row;
	ans.col = col;
	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < col; ++j) {
			ans.data[p(i, j)] = -data[p(i, j)];
		}
	}
	return ans;
}

matrix matrix::operator-(const matrix &m) const {
	if (col != m.col || row != m.row) {
		throw "error size for the matrix to minus";
	}
	if (col == 0 || row == 0) {
		throw "cannot add empty matrix";
	}
	matrix ans;
	ans.data = new fraction[row * col];
	ans.row = row;
	ans.col = col;

	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < col; ++j) {
			ans.data[p(i, j)] = data[p(i, j)] - m.data[p(i, j)];
		}
	}

	return ans;
}

matrix matrix::operator*(const matrix &m) const {
	// 使用最简单的定义计算
	// 时间复杂度 O(n^3)
	// 这样算系数比较小，计算小矩阵乘法速度很快
	if (row == 0 || col == 0 || m.row == 0 || m.col == 0) {
		throw "cannot multiply empty matrix";
	}
	if (col != m.row) {
		// 普通的矩阵乘法前者的列数需要等于后者的行数
		throw "error size for the matrix to multiply";
	}
	matrix ans;
	ans.data =
	    new fraction[row * m.col](); // 根据 ISO C++03 5.3.4
	                                 // 标准，最后加上这个括号可以 new
	                                 // 数组时执行 0 值初始化（默认构造函数）

	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < m.col; ++j) {
			for (int k = 0; k < col; ++k) {
				ans.data[p(i, j)] += data[p(i, k)] * m.data[p(k, j)];
			}
		}
	}

	ans.row = row;
	ans.col = m.col;
	return ans;
}

void matrix::swapRows(int a, int b) {
	for (int i = 0; i < col; i++) {
		fraction temp(data[p(a, i)]);
		data[p(a, i)] = data[p(b, i)];
		data[p(b, i)] = temp;
	}
}

// 转化为上三角矩阵（难点）
int matrix::toUpperTriangularMat(bool step) {
	if (row <= 0) {
		throw "cannot calculate the upper triangular mat of an empty mat";
	}
	if (row > col) {
		throw "row is greater than col";
	}
	int ans(1);

	// 化简为上三角矩阵
	// l 作为行，i 作为列数
	// 以每一行的第一个数作为因子化简
	for (int i = 1; i < row; ++i) {
		// 第一步：行交换，如果遇到一行前几项为 0，则将这一行放在后面
		// 根据定理，每次交换行时需要将答案取相反数
		int l = i - 1;
		for (; l < row && data[p(l, i - 1)] == 0; l++) {
		}
		if (l != i - 1) {
			swapRows(l, i - 1);
			ans = -ans;
			if (step) {
				cout << "r" << l + 1 << " <-> "
				     << "r" << i << endl;
				cout << *this << endl;
			}
		}
		// 第二步：将每一行乘以相应系数，消去后几行的前几项
		fraction inv(1 / data[p(i - 1, i - 1)]);
		for (int k = i; k < row; k++) {
			fraction numToMul(inv * -data[p(k, i - 1)]);
			if (step) {
				cout << "r" << k + 1 << " + (" << numToMul << ")r" << i << endl;
			}
			for (int j = 0; j < col; ++j) {
				data[p(k, j)] += numToMul * data[p(i - 1, j)];
			}
		}
		if (step) {
			cout << *this << endl;
		}
	}
	return ans;
}

// 转化为下三角矩阵，方法与转化为上三角矩阵一样，只不过正好对称
// 必须先运行转化为上三角矩阵，再运行此函数才有意义
int matrix::toLowerTriangularMat(bool step) {
	if (row <= 0) {
		throw "cannot calculate the upper triangular mat of an empty mat";
	}
	if (row > col) {
		throw "row is greater than col";
	}
	int ans(1);

	for (int i = row - 2; i >= 0; --i) {
		fraction inv(1 / data[p(i + 1, i + 1)]);
		for (int k = 0; k <= i; k++) {
			fraction numToMul(inv * -data[p(k, i + 1)]);
			if (step) {
				cout << "r" << k + 1 << " + (" << numToMul << ")r" << i + 2
				     << endl;
			}
			for (int j = 0; j < col; ++j) {
				data[p(k, j)] += numToMul * data[p(i + 1, j)];
			}
		}
		if (step) {
			cout << *this << endl;
		}
	}
	return ans;
}

// 计算行列式（难点）
fraction matrix::det(bool step) const {
	// 此方法是使用行化简的方法计算的行列式
	// 时间复杂度 O(n^3)
	// 计算行列式绝不能使用定义法计算，那样太太太慢了，具体见报告

	if (step) {
		cout << "Input:\n" << *this << endl;
	}

	// 复制一个新的矩阵，以避免修改对象数据
	matrix a(*this);

	// 转化为上三角矩阵，使用初等行变换，ans 记录符号
	fraction ans = a.toUpperTriangularMat(step);

	// 答案就是矩阵的 trace 的乘积咯，注意系数
	for (int i = 0; i < row; ++i) {
		ans *= a.data[p(i, i)];
	}
	return ans;
}

// 计算矩阵逆（难点）
matrix matrix::inv(bool step) const {
	// 若使用伴随矩阵法，其时间复杂度为 O(n^5)
	// 这里使用行变换法，时间复杂度仅为 O(n^3)
	if (row <= 0 || col <= 0) {
		throw "cannot calculate inverse of an empty matrix";
	}
	if (row != col) {
		throw "not mat square";
	}
	// 创建一个新的增广矩阵 (A, E)
	matrix mat;
	mat.data = new fraction[col * row * 2];
	mat.row = row;
	mat.col = 2 * col;
	for (int r = 0; r < row; ++r) {
		for (int c = 0; c < col; ++c) {
			mat.data[mat.p(r, c)] = data[p(r, c)];
		}
	}
	for (int r = 0; r < row; ++r) {
		mat.data[p(1 + r * 2, r)] = 1;
	}
	if (step) {
		cout << "Let's assume the new Matrix B" << endl;
		cout << mat << endl;
	}

	// 化简为上三角矩阵
	mat.toUpperTriangularMat(step);

	// 检查矩阵的 Rank
	fraction test(1);
	for (int i = 0; i < mat.row; ++i) {
		test *= mat.data[mat.p(i, i)];
	}
	if (test == 0) {
		throw "the determinant is 0";
	}

	// 现在做半侧已经是上三角矩阵，下面将做半侧化简为对角矩阵。
	mat.toLowerTriangularMat(step);

	// 转化为单位矩阵
	for (int i = 0; i < mat.row; ++i) {
		if (mat.data[mat.p(i, i)] != 1) {
			fraction inv(1 / mat.data[mat.p(i, i)]);
			if (step) {
				cout << "r" << i + 1 << " x (" << inv << ")" << endl;
			}
			for (int j = 0; j < mat.col; ++j) {
				mat.data[mat.p(i, j)] *= inv;
			}
		}
	}

	if (step) {
		cout << mat << endl;
	}

	// 取结果为右半边的矩阵
	matrix ans;
	ans.col = col;
	ans.row = row;
	ans.data = new fraction[col * row];
	for (int i = 0; i < row; ++i) {
		for (int j = 0; j < col; ++j) {
			ans.data[p(i, j)] = mat.data[p(i * 2 + 1, j)];
		}
	}
	return ans;
}

// 计算增广矩阵的逆（难点）
matrix matrix::inv2(const matrix &c, bool step) const {
	// 若使用伴随矩阵法，其时间复杂度为 O(n^5)
	// 这里使用行变换法，时间复杂度仅为 O(n^3)
	if (row <= 0 || col <= 0) {
		throw "cannot calculate inverse of an empty matrix";
	}
	if (row != col || row != c.row) {
		throw "not mat square";
	}
	fraction det1 = det();
	if (det1 == 0) {
		throw "the determinant is 0";
	}
	// 创建一个新的增广矩阵 (A, C)
	matrix mat;
	mat.data = new fraction[(col + c.col) * row];
	mat.row = row;
	mat.col = c.col + col;
	for (int r = 0; r < row; ++r) {
		for (int c1 = 0; c1 < col; ++c1) {
			mat.data[mat.p(r, c1)] = data[p(r, c1)];
		}
	}
	// 设置右半边矩阵
	for (int r = 0; r < row; ++r) {
		for (int c1 = 0; c1 < c.col; ++c1) {
			mat.data[mat.p(r, col + c1)] = c.data[c.p(r, c1)];
		}
	}
	if (step) {
		cout << "Let's assume the new Matrix B" << endl;
		cout << mat << endl;
	}

	// 化简为上三角矩阵
	mat.toUpperTriangularMat(step);

	// 检查矩阵的 Rank
	fraction test(1);
	for (int i = 0; i < mat.row; ++i) {
		test *= mat.data[mat.p(i, i)];
	}
	if (test == 0) {
		throw "the determinant is 0";
	}

	// 现在做半侧已经是上三角矩阵，下面将做半侧化简为对角矩阵。
	mat.toLowerTriangularMat(step);

	// 转化为单位矩阵
	for (int i = 0; i < mat.row; ++i) {
		if (mat.data[mat.p(i, i)] != 1) {
			fraction inv(1 / mat.data[mat.p(i, i)]);
			if (step) {
				cout << "r" << i + 1 << " x (" << inv << ")" << endl;
			}
			for (int j = 0; j < mat.col; ++j) {
				mat.data[mat.p(i, j)] *= inv;
			}
		}
	}

	if (step) {
		cout << mat << endl;
	}

	// 取结果为右半边的矩阵
	matrix ans;
	ans.col = c.col;
	ans.row = c.row;
	ans.data = new fraction[c.col * c.row];
	for (int i = 0; i < c.row; ++i) {
		for (int j = 0; j < c.col; ++j) {
			ans.data[c.p(i, j)] = mat.data[mat.p(i, col + j)];
		}
	}
	return ans;
}

matrix matrix::transpose() const {
	matrix ans;
	ans.row = row;
	ans.col = col;
	ans.data = new fraction[row*col];
	for (int i = 0; i < col; ++i) {
		for (int j = 0; j < row; ++j) {
			ans.data[p(j, i)] = data[p(i, j)];
		}
	}
	return ans;
}
